assets/src/edit-story/components/panels/textStyle/font.js

```diff
@@ -47,16 +47,53 @@ function FontControls({ selectedElements, pushUpdate }) {
   const fontFamily = getCommonValue(selectedElements, 'fontFamily');
   const fontSize = getCommonValue(selectedElements, 'fontSize');
   const fontWeight = getCommonValue(selectedElements, 'fontWeight');
+  const timeout = useRef(null);
 
   const {
     state: { fonts },
-    actions: { getFontWeight, getFontFallback },
+    actions: { maybeEnqueueFontStyle, getFontWeight, getFontFallback },
   } = useFont();
   const fontWeights = useMemo(() => getFontWeight(fontFamily), [
     getFontWeight,
     fontFamily,
   ]);
 
+  const loadFont = useCallback(
+    (family, callback) => {
+      // 0. Check browser support to load fonts using JavaScript
+      if (document?.fonts) {
+        const font = `0 ${family}`;
+
+        // 1. Check if font family was not loaded yet
+        if (!document.fonts.check(font)) {
+          // 2. Try to inject <link /> resource of the font family
+          maybeEnqueueFontStyle(family);
+          // 3. Schedule a re-check since we need to wait the <link /> injection
+          timeout.current = setTimeout(async () => {
+            // 4. Load the new injected font in the client
+            await document.fonts.load(font);
+            // 5. Re-check if font was loaded
+            if (document.fonts.check(font)) {
+              return callback();
+            }
+            // 6. Try again if font still is not loaded
+            return loadFont(family, callback);
```

Comment:
This should not happen indefinitely. If there's some network error, nothing happens here continuously. And then selecting multiple fonts from dropdown will create multiple indefinite loops of `setTimeout`'s.

Reply:
I don't understand, why does it need to happen even once? Doesn't fonts.load() give us what we want?

Reply:
In the step 2, the <link>  is injected dynamically when it doesn’t exist in the DOM,, which means that the font source will not be immediately available to load (reflow, HTTP request, etc), so we need to re-check (fonts.load not is reactive as expected, and that’s why the recursive approach is needed). When the <link> is already on DOM, fonts.load works like you said, but that’s not our case. Anyway, what Morten said is a real problem that we should cover, its no so easy to fix that since track errors on document.head.appendChild (which report network errors for this case) not is so trivial as it should be.

Reply:
Can we wait until step 2 is done before proceeding? Otherwise this is rather dangerous. Also, structurally it'd be cleaner:

Step 1: ensure that font-face declaration is loaded.
Step 2: ensure that the correct font variant is loaded.

If not it's fine. But we still likely need to draw a line between font-face-set loading and the actual font loading.

Reply:
Sure! I will take a look on this way.